x-pack/plugins/security/server/lib/authentication/authenticator.ts

```diff
@@ -252,46 +245,46 @@ class Authenticator {
   /**
    * Extracts session value for the specified request. Under the hood it can
    * clear session if it belongs to the provider that is not available.
-   * @param {Hapi.Request} request Request to extract session value for.
-   * @returns {Promise.<Object|null>}
-   * @private
+   * @param request Request to extract session value for.
    */
-  async _getSessionValue(request) {
-    let sessionValue = await this._session.get(request);
+  private async getSessionValue(request: Request) {
+    let sessionValue = await this.session.get<ProviderSession>(request);
 
     // If for some reason we have a session stored for the provider that is not available
     // (e.g. when user was logged in with one provider, but then configuration has changed
     // and that provider is no longer available), then we should clear session entirely.
-    if (sessionValue && !this._providers.has(sessionValue.provider)) {
-      await this._session.clear(request);
+    if (sessionValue && !this.providers.has(sessionValue.provider)) {
+      await this.session.clear(request);
       sessionValue = null;
     }
 
     return sessionValue;
   }
 }
 
-export async function initAuthenticator(server) {
+export async function initAuthenticator(server: Legacy.Server) {
   const session = await Session.create(server);
   const authScope = new AuthScopeService();
   const authenticator = new Authenticator(server, authScope, session);
 
   const loginAttempts = new WeakMap();
-  server.decorate('request', 'loginAttempt', function () {
+  server.decorate('request', 'loginAttempt', function(this: Request) {
     const request = this;
     if (!loginAttempts.has(request)) {
       loginAttempts.set(request, new LoginAttempt());
     }
     return loginAttempts.get(request);
   });
 
-  server.expose('authenticate', (request) => authenticator.authenticate(request));
-  server.expose('deauthenticate', (request) => authenticator.deauthenticate(request));
-  server.expose('registerAuthScopeGetter', (scopeExtender) => authScope.registerGetter(scopeExtender));
+  server.expose('authenticate', (request: Request) => authenticator.authenticate(request));
+  server.expose('deauthenticate', (request: Request) => authenticator.deauthenticate(request));
+  server.expose('registerAuthScopeGetter', (scopeExtender: ScopesGetter) =>
+    authScope.registerGetter(scopeExtender)
+  );
 
-  server.expose('isAuthenticated', async (request) => {
+  server.expose('isAuthenticated', async (request: Request) => {
     try {
-      await server.plugins.security.getUser(request);
+      await (server.plugins as Record<string, any>).security.getUser(request);
```

Comment:
Can we define the security plugin [here](https://github.com/elastic/kibana/blob/ed795d28ee780928b307f2271ee6acd9404d7485/x-pack/typings/hapi.d.ts) to avoid the cast to `any`?

I have an example of doing this for the Spaces NP migration: 
[Security Plugin Definition](https://github.com/elastic/kibana/pull/35429/files?file-filters%5B%5D=.ts#diff-23b842ee698fbba6c0e6426b16a7ccea)
[Updated Typings](https://github.com/elastic/kibana/pull/35429/files?file-filters%5B%5D=.ts#diff-c72935f68daa22c5cb9553c33f1f1a02)

Reply:
Sure, but it should be `security?: SecurityPlugin;` I guess since it's an optional plugin, so we'll have to use `!` in `server.plugins.security!.`, still better than casting to `any` - will do!